workloads/librispeech/librispeech_jax/models.py

```diff
@@ -0,0 +1,257 @@
+  Returns:
+    An ndarray with the flipped inputs.
+  """
+  # Note: since this function is vmapped, the code below is effectively for
+  # a single example.
+  max_length = inputs.shape[0]
+  return jnp.flip(jnp.roll(inputs, max_length - lengths, axis=0), axis=0)
+
+
+class SimpleLSTM(nn.Module):
+  """A simple unidirectional LSTM."""
+
+  @functools.partial(
+      nn.transforms.scan,
+      variable_broadcast='params',
+      in_axes=1, out_axes=1,
+      split_rngs={'params': False})
+  @nn.compact
+  def __call__(self, carry, x):
+    return nn.OptimizedLSTMCell()(carry, x)
+
+  @staticmethod
+  def initialize_carry(batch_dims, hidden_size):
+    # Use fixed random key since default state init fn is just zeros.
+    return nn.OptimizedLSTMCell.initialize_carry(
+        jax.random.PRNGKey(0), batch_dims, hidden_size)
+
+
+class SimpleBiLSTM(nn.Module):
+  """A simple bi-directional LSTM."""
+  hidden_size: int
+
+  def setup(self):
+    self.forward_lstm = SimpleLSTM()
+    self.backward_lstm = SimpleLSTM()
+
+  def __call__(self, embedded_inputs, lengths):
+    batch_size = embedded_inputs.shape[0]
+
+    # Forward LSTM.
+    initial_state = SimpleLSTM.initialize_carry((batch_size,), self.hidden_size)
+    _, forward_outputs = self.forward_lstm(initial_state, embedded_inputs)
+
+    # Backward LSTM.
+    reversed_inputs = flip_sequences(embedded_inputs, lengths)
+    initial_state = SimpleLSTM.initialize_carry((batch_size,), self.hidden_size)
+    _, backward_outputs = self.backward_lstm(initial_state, reversed_inputs)
+    backward_outputs = flip_sequences(backward_outputs, lengths)
+
+    # Concatenate the forward and backward representations.
+    outputs = jnp.concatenate([forward_outputs, backward_outputs], -1)
+    return outputs
+
+
+class BatchRNN(nn.Module):
+
+  input_size: int
+  hidden_size: int
+  use_batch_norm: bool = True
+  train: bool = False
+
+  def setup(self):
+    if self.use_batch_norm:
+      self.batch_norm = SequenceWise(nn.BatchNorm(use_running_average=True))
+    else:
+      self.batch_norm = None
+    self.rnn = SimpleBiLSTM(self.hidden_size)
+
+  def __call__(self, inputs, lengths, training=False):
+    if self.batch_norm is not None:
+      inputs = self.batch_norm(inputs)
+    inputs = inputs.transpose(1, 0, 2)
+    inputs = self.rnn(inputs, lengths)
+    inputs = inputs.transpose(1, 0, 2)
+    # (TxNxH*2) -> (TxNxH) by sum
+    inputs = inputs.reshape(inputs.shape[0], inputs.shape[1], -1).sum(2)
+    return inputs
+
+
+def hard_tanh(x, min_value=-1., max_value=1.):
+
+  return jnp.where(x > max_value, max_value, jnp.where(x < min_value, min_value, x))
+
+
+def get_seq_lens(input_length, conv_seq_module):
+    """Get a 1D tensor or variable containing the size sequences that will be output by the network.
+
+    Args:
+      input_length: 1D Tensor
+
+    Returns:
+      1D Tensor scaled by model
+    """
+    seq_len = input_length
+    for m in conv_seq_module:
+      if isinstance(m, nn.Conv):
```

Comment:
is there ever an instance where this is false?

Reply:
yes

Reply:
`conv_seq_module` also contains `BatchNorm` and `hard_tanh`.